Add new setting to disable persistent tasks allocations

tlrx opened the pull request:
This commit adds a new setting `cluster.persistent_tasks.allocation.enable` that can be used to enable or disable the allocation of persistent tasks. 

The setting accepts the values `all` (default) or `none`. When set to none, the persistent tasks that are created (or that must be reassigned) won't be assigned to a node but will reside in the cluster state with
a `null` executor node and a reason describing why it is not assigned:

```
"assignment" : {
  "executor_node" : null,
  "explanation" : "persistent task [foo/bar] cannot be assigned [no persistent task assignments are allowed due to cluster settings]"
}
```

This commit adds some AssignmentDecider and AssignmentDecision classes that are similar to the deciders/decisions classes used for shard allocations. It also adds an EnableAssignmentDecider decider that is responsible for registering the new cluster setting.

The setting is checked when executing the `PersistentTasksClusterService#createAssignment()` method, before delegating the creation of the assignment to the PersistentTasksExecutor. As a follow up, I think that PersistentTasksExecutor could expose a `getTaskAllocationDecider` method that would be used by the persistent task cluster service to build a list of nodes that could execute the new task. Then this list of assignable nodes could be passed again to the executor to select the exact node to execute the task.

elasticmachine commented:
Pinging @elastic/es-distributed

imotov commented on server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java:
```diff
@@ -0,0 +1,78 @@
+ */
+public final class AssignmentDecision {
+
+    public static final AssignmentDecision ALWAYS = new AssignmentDecision(Type.YES);
```
That feels like premature optimization.

bleskes replied:
+1 . People can just use yes? I'm not sure what the difference is?

tlrx replied:
Right, it's not required.

tlrx commented:
> Could you open an issue for the follow-up that you proposed at the end of your PR so it wouldn't get lost? Without such follow-up this PR doesn't make a lot of sense. Ideally, I would have preferred for this follow-up to happen first and this change to follow, but it might be not possible if we need this for 6.3.

Yes, the follow up is a bigger change and we'd like the setting to be added in 6.3, thus the order of pull requests. Thanks for your review Igor.

bleskes commented on server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java:
```diff
@@ -45,12 +48,14 @@
 
     private final ClusterService clusterService;
     private final PersistentTasksExecutorRegistry registry;
+    private final AssignmentDecider<PersistentTaskParams> decider;
```
can we just hard code this for now to EnableAssignmentDecider? we can generalize when it becomes relevant (and know what we want exactly)

bleskes commented on server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecider.java:
```diff
@@ -0,0 +1,47 @@
+ * {@link AssignmentDecider} are used to decide if a persistent
+ * task can be assigned to a node.
+ */
+public abstract class AssignmentDecider<Params extends PersistentTaskParams> extends AbstractComponent {
```
I would prefer this to be added at the same time we support multiple deciders.

bleskes commented on server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java:
```diff
@@ -0,0 +1,78 @@
+
+    public static final AssignmentDecision ALWAYS = new AssignmentDecision(Type.YES);
+    public static final AssignmentDecision YES = new AssignmentDecision(Type.YES);
+    public static final AssignmentDecision NO = new AssignmentDecision(Type.NO);
```
I'm not sure we want this? it means people can say no without giving a reason.

tlrx replied:
Right

bleskes commented on server/src/main/java/org/elasticsearch/persistent/decider/EnableAssignmentDecider.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ * @see Allocation
+ */
+public class EnableAssignmentDecider<Params extends PersistentTaskParams> extends AssignmentDecider<Params> {
```
Why is this generic?

tlrx commented:
Thanks @imotov and @bleskes and sorry for all the premature stuff.

@bleskes I addressed your comments and also added an integration test (but not a QA one), please let me know what you think.

tlrx commented:
Thanks @bleskes
